### ISSUE TITLE
Illegal Access running under Java 11 and Spring Boot

### ISSUE THREAD

**Issue by minborg** [OpenHFT/Chronicle-Core#110]
When running Datagrid under Spring and Java 8, the application starts flawlessly:

```
   _____ _                     _      _        _____        _                   _     _ 
  / ____| |                   (_)    | |      |  __ \      | |                 (_)   | |
 | |    | |__  _ __ ___  _ __  _  ___| | ___  | |  | | __ _| |_ __ _  __ _ _ __ _  __| |
 | |    | '_ \| '__/ _ \| '_ \| |/ __| |/ _ \ | |  | |/ _` | __/ _` |/ _` | '__| |/ _` |
 | |____| | | | | | (_) | | | | | (__| |  __/ | |__| | (_| | || (_| | (_| | |  | | (_| |
  \_____|_| |_|_|  \___/|_| |_|_|\___|_|\___| |_____/ \__,_|\__\__,_|\__, |_|  |_|\__,_|
                                                                      __/ |             
:: Chronicle Datagrid ::      (2.19.8-SNAPSHOT)                      |___/              
Server ID:   1
Configuration file: /Users/pemi/git/chronicle/customer/chronicle-datagrid-sdk-test/datagrid.yaml
Running under Java(TM) SE Runtime Environment 1.8.0_191-b12 with 16 processors reported.
Listening for incoming connections on port 0.0.0.0:9090
Press CTRL+C to kill the Chronicle Datagrid Server process.

```

However, using the same setup but running under Java 11 will produce an illegal access:

```
2020-05-08 12:54:11.797  WARN 35412 --- [           main] s.c.d.ServerLauncher                     : The file /Users/pemi/git/chronicle/customer/chronicle-datagrid-sdk-test/datagrid.yaml does not exist.
   _____ _                     _      _        _____        _                   _     _ 
  / ____| |                   (_)    | |      |  __ \      | |                 (_)   | |
 | |    | |__  _ __ ___  _ __  _  ___| | ___  | |  | | __ _| |_ __ _  __ _ _ __ _  __| |
 | |    | '_ \| '__/ _ \| '_ \| |/ __| |/ _ \ | |  | |/ _` | __/ _` |/ _` | '__| |/ _` |
 | |____| | | | | | (_) | | | | | (__| |  __/ | |__| | (_| | || (_| | (_| | |  | | (_| |
  \_____|_| |_|_|  \___/|_| |_|_|\___|_|\___| |_____/ \__,_|\__\__,_|\__, |_|  |_|\__,_|
                                                                      __/ |             
:: Chronicle Datagrid ::      (2.19.8-SNAPSHOT)                      |___/              
Server ID:   1
Configuration file: /Users/pemi/git/chronicle/customer/chronicle-datagrid-sdk-test/datagrid.yaml
Running under OpenJDK Runtime Environment 11.0.6+8-b765.40 with 16 processors reported.
2020-05-08 12:54:11.864  WARN 35412 --- [           main] n.o.c.c.c.CleanerServiceLocator          : Error while trying to load service providers

java.util.ServiceConfigurationError: net.openhft.chronicle.core.cleaner.spi.ByteBufferCleanerService: Provider net.openhft.chronicle.core.cleaner.impl.reflect.ReflectionBasedByteBufferCleanerService could not be instantiated
	at java.util.ServiceLoader.fail(ServiceLoader.java:581) ~[?:?]
	at java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:803) ~[?:?]
	at java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:721) ~[?:?]
	at java.util.ServiceLoader$3.next(ServiceLoader.java:1394) ~[?:?]
	at net.openhft.chronicle.core.cleaner.CleanerServiceLocator.cleanerService(CleanerServiceLocator.java:27) [chronicle-core-2.19.6.jar:?]
	at net.openhft.chronicle.bytes.NativeBytesStore.<clinit>(NativeBytesStore.java:44) [chronicle-bytes-2.19.7.jar:?]
	at net.openhft.chronicle.bytes.BytesStore.from(BytesStore.java:52) [chronicle-bytes-2.19.7.jar:?]
	at net.openhft.chronicle.wire.WireType.<clinit>(WireType.java:328) [chronicle-wire-2.19.11.jar:?]
	at software.chronicle.datagrid.internal.ServerFactory.createAssetTree(ServerFactory.java:76) [chronicle-datagrid-2.19.8-SNAPSHOT.jar:2.19.8-SNAPSHOT]
	at software.chronicle.datagrid.internal.ServerFactory.datagridMain(ServerFactory.java:61) [chronicle-datagrid-2.19.8-SNAPSHOT.jar:2.19.8-SNAPSHOT]
	at software.chronicle.datagrid.internal.VanillaServerBuilder.build(VanillaServerBuilder.java:62) [chronicle-datagrid-2.19.8-SNAPSHOT.jar:2.19.8-SNAPSHOT]
	at software.chronicle.datagrid.ServerLauncher.run(ServerLauncher.java:114) [chronicle-datagrid-2.19.8-SNAPSHOT.jar:2.19.8-SNAPSHOT]
	at com.okcoin.chronicle.sender.Server.run(Server.java:31) [classes/:?]
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:775) [spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:765) [spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:322) [spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) [spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215) [spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
	at com.okcoin.chronicle.sender.Application.main(Application.java:11) [classes/:?]
Caused by: java.lang.ExceptionInInitializerError
	at net.openhft.chronicle.core.cleaner.impl.reflect.ReflectionBasedByteBufferCleanerService.<clinit>(ReflectionBasedByteBufferCleanerService.java:28) ~[chronicle-core-2.19.6.jar:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:490) ~[?:?]
	at java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:779) ~[?:?]
	... 17 more
Caused by: java.lang.IllegalAccessException: symbolic reference class is not accessible: class jdk.internal.ref.Cleaner, from net.openhft.chronicle.core.cleaner.impl.reflect.ReflectionBasedByteBufferCleanerService (unnamed module @7ed9499e)
	at java.lang.invoke.MemberName.makeAccessException(MemberName.java:942) ~[?:?]
	at java.lang.invoke.MethodHandles$Lookup.checkSymbolicClass(MethodHandles.java:2054) ~[?:?]
	at java.lang.invoke.MethodHandles$Lookup.resolveOrFail(MethodHandles.java:2026) ~[?:?]
	at java.lang.invoke.MethodHandles$Lookup.findVirtual(MethodHandles.java:1194) ~[?:?]
	at net.openhft.chronicle.core.cleaner.impl.reflect.ReflectionBasedByteBufferCleanerService.<clinit>(ReflectionBasedByteBufferCleanerService.java:26) ~[chronicle-core-2.19.6.jar:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:490) ~[?:?]
	at java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:779) ~[?:?]
	... 17 more

```

**Comment by dpisklov** [comment]
This has been fixed in latest core

**Comment by minborg** [comment]
Unfortunately not. I am running 2.19.7-SNAPSHOT:

```
public final class ReflectionBasedByteBufferCleanerService implements ByteBufferCleanerService {
    private static final String JDK8_CLEANER_CLASS_NAME = "sun.misc.Cleaner";
    private static final String JDK9_CLEANER_CLASS_NAME = "jdk.internal.ref.Cleaner";

    private static final MethodHandle cleanerMethod, cleanMethod;

    static {
        MethodHandles.Lookup lookup = MethodHandles.lookup();
        final String cleanerClassname = Jvm.isJava9Plus() ?
                JDK9_CLEANER_CLASS_NAME : JDK8_CLEANER_CLASS_NAME;
        try {
            final Class<?> cleanerClass = Class.forName(cleanerClassname);
            cleanerMethod = lookup.findVirtual(DirectBuffer.class, "cleaner", MethodType.methodType(cleanerClass));
            cleanMethod = lookup.findVirtual(cleanerClass, "clean", MethodType.methodType(void.class));
        } catch (NoSuchMethodException | ClassNotFoundException | IllegalAccessException e) {
            throw new ExceptionInInitializerError(e);
        }
    }

    @Override
    public void clean(final ByteBuffer buffer) {
        try {
            Object cleaner = cleanerMethod.invoke((DirectBuffer) buffer);
            cleanMethod.invoke(cleaner);
        } catch (Throwable throwable) {
            Jvm.rethrow(throwable);
        }
    }

    @Override
    public int impact() {
        return SOME_IMPACT;
    }



```

**Comment by minborg** [comment]
If I run the application under Spring, I see the problem. If I run datagrid as a stand-alone app, I do not see it.

**Comment by dpisklov** [comment]
I tried it under java 11 and it worked, may be you've got some dependency
issue and it's picked up wrong version?

**Comment by minborg** [comment]
Indeed I had that problem when I first came across the problem but presumably not now:


```
pemi@Pers-MBP-2 chronicle-datagrid-sdk-test % mvn dependency:tree
[INFO] Scanning for projects...
[INFO] 
[INFO] --------------------< com.okcoin.sdk.test:skd-test >--------------------
[INFO] Building skd-test 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-dependency-plugin:3.1.1:tree (default-cli) @ skd-test ---
[INFO] com.okcoin.sdk.test:skd-test:jar:1.0-SNAPSHOT
[INFO] +- software.chronicle:chronicle-datagrid:jar:2.19.8-SNAPSHOT:compile
[INFO] |  +- net.openhft:chronicle-wire:jar:2.19.11:compile
[INFO] |  +- net.openhft:chronicle-queue:jar:5.19.14:compile
[INFO] |  |  \- commons-cli:commons-cli:jar:1.4:compile
[INFO] |  +- software.chronicle:chronicle-queue-enterprise:jar:2.19.19:compile
[INFO] |  +- net.openhft:chronicle-map:jar:3.19.4:compile
[INFO] |  |  +- net.openhft:chronicle-values:jar:2.19.0:compile
[INFO] |  |  |  \- com.squareup:javapoet:jar:1.5.1:compile
[INFO] |  |  +- net.openhft:chronicle-algorithms:jar:2.19.0:compile
[INFO] |  |  +- net.java.dev.jna:jna:jar:4.5.2:compile
[INFO] |  |  +- net.java.dev.jna:jna-platform:jar:4.5.2:compile
[INFO] |  |  +- com.thoughtworks.xstream:xstream:jar:1.4.9:compile
[INFO] |  |  |  +- xmlpull:xmlpull:jar:1.1.3.1:compile
[INFO] |  |  |  \- xpp3:xpp3_min:jar:1.1.4c:compile
[INFO] |  |  +- org.codehaus.jettison:jettison:jar:1.3.8:compile
[INFO] |  |  |  \- stax:stax-api:jar:1.0.1:compile
[INFO] |  |  \- org.ops4j.pax.url:pax-url-aether:jar:2.4.5:compile
[INFO] |  |     \- org.slf4j:jcl-over-slf4j:jar:1.7.29:compile
[INFO] |  +- net.openhft:compiler:jar:2.3.5:compile
[INFO] |  |  \- com.intellij:annotations:jar:12.0:compile
[INFO] |  +- software.chronicle:chronicle-map-enterprise:jar:2.19.1:compile
[INFO] |  |  \- net.openhft:affinity:jar:3.2.3:compile
[INFO] |  +- net.openhft:chronicle-network:jar:2.19.8-SNAPSHOT:compile
[INFO] |  +- org.slf4j:slf4j-api:jar:1.7.29:compile
[INFO] |  +- net.openhft:chronicle-bytes:jar:2.19.7:compile
[INFO] |  |  \- org.jetbrains:annotations:jar:19.0.0:compile
[INFO] |  +- net.openhft:chronicle-threads:jar:2.19.6:compile
[INFO] |  +- net.openhft:chronicle-core:jar:2.19.7-SNAPSHOT:compile
[INFO] |  \- net.sf.trove4j:core:jar:3.1.0:compile
[INFO] +- org.projectlombok:lombok:jar:1.18.10:provided
[INFO] +- org.springframework.boot:spring-boot-starter-log4j2:jar:2.2.1.RELEASE:compile
[INFO] |  +- org.apache.logging.log4j:log4j-slf4j-impl:jar:2.12.1:compile
[INFO] |  |  \- org.apache.logging.log4j:log4j-api:jar:2.12.1:compile
[INFO] |  +- org.apache.logging.log4j:log4j-core:jar:2.12.1:compile
[INFO] |  +- org.apache.logging.log4j:log4j-jul:jar:2.12.1:compile
[INFO] |  \- org.slf4j:jul-to-slf4j:jar:1.7.29:compile
[INFO] +- org.springframework.boot:spring-boot-starter:jar:2.2.1.RELEASE:compile
[INFO] |  +- org.springframework.boot:spring-boot:jar:2.2.1.RELEASE:compile
[INFO] |  |  \- org.springframework:spring-context:jar:5.2.1.RELEASE:compile
[INFO] |  +- org.springframework.boot:spring-boot-autoconfigure:jar:2.2.1.RELEASE:compile
[INFO] |  +- jakarta.annotation:jakarta.annotation-api:jar:1.3.5:compile
[INFO] |  +- org.springframework:spring-core:jar:5.2.1.RELEASE:compile
[INFO] |  |  \- org.springframework:spring-jcl:jar:5.2.1.RELEASE:compile
[INFO] |  \- org.yaml:snakeyaml:jar:1.25:runtime
[INFO] \- org.springframework.boot:spring-boot-starter-web:jar:2.1.7.RELEASE:compile
[INFO]    +- org.springframework.boot:spring-boot-starter-json:jar:2.2.1.RELEASE:compile
[INFO]    |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.10.0:compile
[INFO]    |  |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.10.0:compile
[INFO]    |  |  \- com.fasterxml.jackson.core:jackson-core:jar:2.10.0:compile
[INFO]    |  +- com.fasterxml.jackson.datatype:jackson-datatype-jdk8:jar:2.10.0:compile
[INFO]    |  +- com.fasterxml.jackson.datatype:jackson-datatype-jsr310:jar:2.10.0:compile
[INFO]    |  \- com.fasterxml.jackson.module:jackson-module-parameter-names:jar:2.10.0:compile
[INFO]    +- org.hibernate.validator:hibernate-validator:jar:6.0.18.Final:compile
[INFO]    |  +- javax.validation:validation-api:jar:2.0.1.Final:compile
[INFO]    |  +- org.jboss.logging:jboss-logging:jar:3.4.1.Final:compile
[INFO]    |  \- com.fasterxml:classmate:jar:1.5.1:compile
[INFO]    +- org.springframework:spring-web:jar:5.2.1.RELEASE:compile
[INFO]    |  \- org.springframework:spring-beans:jar:5.2.1.RELEASE:compile
[INFO]    \- org.springframework:spring-webmvc:jar:5.2.1.RELEASE:compile
[INFO]       +- org.springframework:spring-aop:jar:5.2.1.RELEASE:compile
[INFO]       \- org.springframework:spring-expression:jar:5.2.1.RELEASE:compile

```

**Comment by minborg** [comment]
Note that the problem only manifested itself under Spring and is related to the ServiceLoader

**Comment by minborg** [comment]
`export MAVEN_OPTS="--illegal-access=permit --add-exports java.base/jdk.internal.ref=ALL- UNNAMED"` makes the problem go away.

**Comment by minborg** [comment]
Unless no one else is inclined to take on this, I will fix the issue.

**Comment by minborg** [comment]
Related to https://github.com/OpenHFT/Chronicle-Network/issues/66

**Comment by RobAustin** [comment]
ok

**Comment by hft-team-city** [comment]
Released in [Chronicle-Core-2.20.102](https://github.com/OpenHFT/Chronicle-Core/releases/chronicle-core-2.20.102), [BOM-2.20.134](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.20.134)
